src/editor/grid.rs

```diff
@@ -0,0 +1,76 @@
+use std::sync::Arc;
+use log::trace;
+
+use super::style::Style;
+
+type GridCell = Option<(String, Option<Arc<Style>>)>;
+
+pub struct CharacterGrid {
+    pub characters: Vec<GridCell>,
+    pub width: u64,
+    pub height: u64,
+    pub should_clear: bool,
+
+    dirty: Vec<bool>,
+}
+
+impl CharacterGrid {
+    pub fn new() -> CharacterGrid {
+        CharacterGrid {
+            characters: vec![],
+            dirty: vec![],
+            width: 0,
+            height: 0,
+            should_clear: true,
+        }
+    }
+
+    pub fn resize(&mut self, new_size: (u64, u64)) {
+        trace!("Editor resized");
+        self.width = new_size.0;
+        self.height = new_size.1;
+        self.clear();
+    }
+
+    pub fn clear(&mut self) {
+        trace!("Editor cleared");
+        let cell_count = (self.width * self.height) as usize;
+        self.characters = vec![None; cell_count];
+        self.dirty = vec![true; cell_count];
+        self.should_clear = true;
+    }
+
+    pub fn cell_index(&self, x: u64, y: u64) -> Option<usize> {
```

Comment:
Instead of this, couldn't we have a method to return a reference for the cell? From the code, I see the following pattern:

1. Get the index.
2. Use the index to get the cell.

This doesn't abstract the underlying storage of the cells and is very repetitive.

Reply:
Yep, returning an `Option<&'a mut GridCell>` could be a better interface. I'll look into that. Thanks for the suggestion :+1:

Reply:
I suggest making two methods, one for `&GridCell` and another for `&mut GridCell`, this way we can write functions that don't need mutable access to `CharacterGrid`.

Reply:
Yep, that was my plan